### PR TITLE
Keep track of the best model's path saved by ModelCheckpoint

### PR DESCRIPTION
**Pull request by kepler**
# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?   
- [ ] Did you write any new necessary tests?  
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## What does this PR do?
Add an additional attribute to `ModelCheckpoint` to keep track of the best model's path. Currently, only the best metric value is directly tracked.

When training finishes, if we want to know where the best model is, we need to either do:
```
op = min if checkpoint_callback.mode == 'min' else max
best_model_path = op(
    checkpoint_callback.best_k_models,
    key=checkpoint_callback.best_k_models.get
)
```
or
```
best_model_path = [
    path
    for path, metric in checkpoint_callback.best_k_models.items()
    if metric == checkpoint_callback.best
][0]
```
which are both somewhat cumbersome.

Since the best value is already tracked, this PR simply tracks the saved path as well.

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Glad to make PTL even more practical.

### CODE REVIEW

**Comment by williamFalcon** [comment]
@kepler nice! 
Can you add to the docs?

**Comment by kepler** [comment]
Thanks, @williamFalcon.
Not sure where to add any documentation, though. None of the attributes in `ModelCheckpoint`, like `best` and `best_k_models`, are documented. Should I simply add a note in the class' docstring, right before `Args`?

**Comment by kepler** [comment]
@williamFalcon let me know if this is enough.

**Comment by williamFalcon** [comment]
@kepler that's great!
However, can you do a new PR with the changes?

**Comment by kepler** [comment]
@williamFalcon are you sure this is already merged to `master`? I can't see the changes there (hence creating a new PR based on this branch seems convoluted to me).

**Comment by williamFalcon** [comment]
@kepler you're right! sorry about that. let's merge this PR then :)

**Comment by kepler** [comment]
@williamFalcon no problem.

I have to say I don't understand why the tests failed (for Ubuntu and MacOS but passed for Windows). Could you please rerun them, just in case? Otherwise, I'll trigger it with a small commit (in about two hours).

**Comment by justusschock** [comment]
https://github.com/PyTorchLightning/pytorch-lightning/pull/1799/checks?check_run_id=712909136 they don't pass on windows. For windows this is just not correctly indicated.

On linux and macos, you have a print statement in your doctests which produce some output although no output is expected and thus the tests fail

**Comment by kepler** [comment]
So it was indeed a problem with the doctest (totally missed it). Thanks, @justusschock.

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 1]
```diff
@@ -89,6 +89,7 @@ class ModelCheckpoint(Callback):
         >>> # model = ...
         >>> # trainer.fit(model)
         >>> print(checkpoint_callback.best_model)
+        <BLANKLINE>
```
Comment (Borda) [inline pytorch_lightning/callbacks/model_checkpoint.py]:
```suggestion
        >>> checkpoint_callback.best_model
        ''
```

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 2]
```diff
@@ -80,6 +83,14 @@ class ModelCheckpoint(Callback):
         ...     filepath='my/path/sample-mnist_{epoch:02d}-{val_loss:.2f}'
         ... )
 
+        # retrieve the best checkpoint after training
+        >>> checkpoint_callback = ModelCheckpoint(filepath='my/path/')
+        >>> trainer = Trainer(checkpoint_callback=checkpoint_callback)
+        >>> # model = ...
+        >>> # trainer.fit(model)
+        >>> print(checkpoint_callback.best_model)
+        <BLANKLINE>
```
Comment (Borda) [inline pytorch_lightning/callbacks/model_checkpoint.py:92]:
```suggestion
        >>> checkpoint_callback.best_model
        ''
```

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 3]
```diff
@@ -80,6 +83,14 @@ class ModelCheckpoint(Callback):
         ...     filepath='my/path/sample-mnist_{epoch:02d}-{val_loss:.2f}'
         ... )
 
+        # retrieve the best checkpoint after training
+        >>> checkpoint_callback = ModelCheckpoint(filepath='my/path/')
+        >>> trainer = Trainer(checkpoint_callback=checkpoint_callback)
+        >>> # model = ...
```
Comment (Borda) [inline pytorch_lightning/callbacks/model_checkpoint.py:89]: what is the purpose here, the model and fir is commented

**Comment by Borda** [comment]
> I have to say I don't understand why the tests failed (for Ubuntu and MacOS but passed for Windows). Could you please rerun them, just in case? Otherwise, I'll trigger it with a small commit (in about two hours).

Win is pure and not have implemented all functions properly...

File: pytorch_lightning/callbacks/model_checkpoint.py (hunk 3 above)
Comment (kepler) [reply pytorch_lightning/callbacks/model_checkpoint.py:89]: To better show how it's used without actually instantiating a model. If there's a mockup for a `LightningModule`, we can use it, but I couldn't find any.

Comment (Borda) [reply pytorch_lightning/callbacks/model_checkpoint.py:89]: sure but then show is just as code example not as doc test... just remove `>>>` for these 5 lines (this block, keep the above)

**Comment by kepler** [comment]
@Borda This should be ready now.

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 4]
```diff
@@ -80,6 +83,13 @@ class ModelCheckpoint(Callback):
         ...     filepath='my/path/sample-mnist_{epoch:02d}-{val_loss:.2f}'
         ... )
 
+        # retrieve the best checkpoint after training
+        checkpoint_callback = ModelCheckpoint(filepath='my/path/')
+        trainer = Trainer(checkpoint_callback=checkpoint_callback)
+        # model = ...
+        # trainer.fit(model)
```
Comment (Borda) [inline pytorch_lightning/callbacks/model_checkpoint.py:90]:
```suggestion
        model = ...
        trainer.fit(model)
```

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 5]
```diff
@@ -265,7 +276,8 @@ def _do_check_save(self, filepath, current, epoch):
             self.kth_value = self.best_k_models[self.kth_best_model]
 
         _op = min if self.mode == 'min' else max
-        self.best = _op(self.best_k_models.values())
+        self.best_model = _op(self.best_k_models, key=self.best_k_models.get)
+        self.best = self.best_k_models[self.best_model]
```
Comment (Borda) [inline pytorch_lightning/callbacks/model_checkpoint.py:280]: here the best means `best_score`? otherwise, it is a bit confusing to have `best` and `best_model`
pls add types to the init :rabbit:

Comment (kepler) [reply pytorch_lightning/callbacks/model_checkpoint.py:280]: Yes, `self.best` is the best "monitored quantity". But that's existing naming. Changing it will break backwards compatibility. If that's OK, I can surely rename it. Otherwise, `best_model` could be `best_model_path`. I used `best_model` to keep consistency with the existing `kth_best_model` attribute.

As for type hints, there's no changes to the arguments. Where specifically would I add them?

Comment (Borda) [reply pytorch_lightning/callbacks/model_checkpoint.py:280]: @PyTorchLightning/core-contributors are we fine to rename `best` >> `best_score`?

Comment (jeremyjordan) [reply pytorch_lightning/callbacks/model_checkpoint.py:280]: i would prefer it (`best_score`), but if we do that we need to provide our users with a utility to upgrade their checkpoints (basically switch `best` to `best_score`) and temporarily have a warning which catches old checkpoints at load time and alerts users of the utility to convert. it's a very simple utility but we should provide it for our users if we move forward with this.

Comment (Borda) [reply pytorch_lightning/callbacks/model_checkpoint.py:280]: @jeremyjordan you mean to keep compatible with past saved checkpoint which contains `best`? That shall be simple, in loading, we will map `best` >> `best_score`... and yes we shall wrap the deprecated `best` with some warning :]

Comment (kepler) [reply pytorch_lightning/callbacks/model_checkpoint.py]: OK, I renamed `best` to `best_model_score` and `best_model` to `best_model_path`, since having `best_score` and `best_model` would still be a bit confusing (IMHO). To keep consistency, I also renamed `kth_best_model` to `kth_best_model_path`.

I added properties for `best` and `kth_best_model` that log a deprecation warning and return the correct value.

When loading a checkpoint, if it's in an old format, the value for `best` is simply assigned to `best_model_score`. In my opinion, adding a warning in this part will not really help the user, as there's not much they can do.

Let me know of any further changes.

File: CHANGELOG.md [hunk 6]
```diff
@@ -26,6 +28,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 - Allow passing model hyperparameters as complete kwarg list ([#1896](https://github.com/PyTorchLightning/pytorch-lightning/pull/1896)) 
 
+- Renamed `ModelCheckpoint`'s attributes `best` to `best_model_score` and `kth_best_model` to `kth_best_model_path` ([#1799](https://github.com/PyTorchLightning/pytorch-lightning/pull/1799))
+
 ### Deprecated
 
 - Dropped official support/testing for older PyTorch versions <1.3 ([#1917](https://github.com/PyTorchLightning/pytorch-lightning/pull/1917))
```
Comment (Borda) [inline CHANGELOG.md:35]:
```suggestion
- Deprecated `ModelCheckpoint`'s attributes `best` and `kth_best_model` ([#1799](https://github.com/PyTorchLightning/pytorch-lightning/pull/1799))

- Dropped official support/testing for older PyTorch versions <1.3 ([#1917](https://github.com/PyTorchLightning/pytorch-lightning/pull/1917))
```

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 7]
```diff
@@ -131,6 +142,18 @@ def __init__(self, filepath: Optional[str] = None, monitor: str = 'val_loss', ve
 
         self.kth_value, self.mode = mode_dict[mode]
 
+    @property
+    def best(self):
+        rank_zero_warn("attribute `best` has been renamed to `best_model_score` since v0.8.0"
+                       " and will be removed in v1.0.0", DeprecationWarning)
```
Comment (Borda) [inline pytorch_lightning/callbacks/model_checkpoint.py:148]:
```suggestion
        rank_zero_warn("Attribute `best` has been renamed to `best_model_score` since v0.8.0"
                       " and will be removed in v0.10.0", DeprecationWarning)
```

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 8]
```diff
@@ -131,6 +142,18 @@ def __init__(self, filepath: Optional[str] = None, monitor: str = 'val_loss', ve
...
+        return self.best_model_score
+
+    @property
+    def kth_best_model(self):
+        rank_zero_warn("attribute `kth_best_model` has been renamed to `kth_best_model_path` since v0.8.0"
+                       " and will be removed in v1.0.0", DeprecationWarning)
```
Comment (Borda) [inline pytorch_lightning/callbacks/model_checkpoint.py:154]:
```suggestion
        rank_zero_warn("Attribute `kth_best_model` has been renamed to `kth_best_model_path` since v0.8.0"
                       " and will be removed in v0.10.0", DeprecationWarning)
```

File: pytorch_lightning/trainer/training_io.py [hunk 9]
```diff
@@ -401,10 +402,14 @@ def restore_training_state(self, checkpoint):
...
             )
 
-        if self.checkpoint_callback is not None and self.checkpoint_callback is not False:
-            self.checkpoint_callback.best = checkpoint['checkpoint_callback_best']
+        if self.checkpoint_callback:
+            if 'checkpoint_callback_best' in checkpoint:
+                self.checkpoint_callback.best_model_score = checkpoint['checkpoint_callback_best']
+            else:
+                self.checkpoint_callback.best_model_score = checkpoint['checkpoint_callback_best_model_score']
+            self.checkpoint_callback.best_model_path = checkpoint['checkpoint_callback_best_model_path']
```
Comment (Borda) [inline pytorch_lightning/trainer/training_io.py:410]:
```suggestion
            self.checkpoint_callback.best_model_path = checkpoint.get('checkpoint_callback_best_model_path')
```
as the old models does not have it

File: pytorch_lightning/trainer/training_io.py [hunk 10]
```diff
@@ -401,10 +402,14 @@ def restore_training_state(self, checkpoint):
                 ' This is probably due to `ModelCheckpoint.save_weights_only` being set to `True`.'
             )
 
-        if self.checkpoint_callback is not None and self.checkpoint_callback is not False:
-            self.checkpoint_callback.best = checkpoint['checkpoint_callback_best']
+        if self.checkpoint_callback:
+            if 'checkpoint_callback_best' in checkpoint:
+                self.checkpoint_callback.best_model_score = checkpoint['checkpoint_callback_best']
+            else:
+                self.checkpoint_callback.best_model_score = checkpoint['checkpoint_callback_best_model_score']
```
Comment (Borda) [inline pytorch_lightning/trainer/training_io.py:409]:
```suggestion
            # try to load the new version and if it missing use the old one            
            self.checkpoint_callback.best_model_score = checkpoint.get('checkpoint_callback_best_model_score', checkpoint.get('checkpoint_callback_best'))
```
not sure if it is not too long line

Comment (kepler) [reply pytorch_lightning/trainer/training_io.py:409]: I think it's less clear. We could keep the `if` and add a comment saying it's for checkpoints older than `0.x.y`.

File: pytorch_lightning/trainer/training_io.py (hunk 9 above)
Comment (kepler) [reply pytorch_lightning/trainer/training_io.py:410]: True, good catch.

File: pytorch_lightning/trainer/training_io.py (hunk 10 above)
Comment (Borda) [reply pytorch_lightning/trainer/training_io.py:409]: ok, then swap the logic, fist we try the new and then if it is missing try to get the old one
also add warning that you are loading old checkpoint?

Comment (kepler) [reply pytorch_lightning/trainer/training_io.py]: Done.
